### PR TITLE
feat(tpcds-benchmarking): Add basic tpcds benchmarking for local testing

### PR DESCRIPTION
**Pull request by raunakab**
# Overview
This PR enables TPC-DS benchmarking on your local computer.

If the TPC-DS parquet data does *not* exist, it will be created for you (inside of `benchmarking/tpcds/data`). You can configure the location of this output via the `--tpcds-gen-folder` argument provided by the script.

## Usage

For running with the native runner, run:

```sh
DAFT_RUNNER=native python -m benchmarking.tpcds --questions "3"
```

For executing with the ray runner, run:

```sh
DAFT_RUNNER=ray python -m benchmarking.tpcds --questions "3"
```

You can also specify different scale-factors (e.g., `--scale-factor 0.5`) and dry-run modes (e.g., `--dry-run`).

### CODE REVIEW

File: benchmarking/tpcds/__main__.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+import daft
+from daft.sql import SQLCatalog
+
+from . import datagen
+
+
+def generate_tpcds_catalog(tpcds_gen_folder: Path):
+    if not tpcds_gen_folder.exists():
+        raise RuntimeError
```
Comment (jaychia) [inline benchmarking/tpcds/__main__.py:14]: Give proper error message

File: benchmarking/tpcds/__main__.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    return SQLCatalog(
+        tables={
+            file.stem: daft.read_parquet(path=str(file))
+            for file in tpcds_gen_folder.iterdir()
+            if file.is_file() and file.suffix == ".parquet"
+        }
```
Comment (jaychia) [inline benchmarking/tpcds/__main__.py:20]: Not part of this PR, but @universalmind303 do you think we should start using `daft.register_named_table` instead so we can unify the catalog experience around that?

File: benchmarking/tpcds/__main__.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    )
+
+
+def run_query(query_index: int, catalog: SQLCatalog) -> str:
+    sql_queries_path = Path("benchmarking") / "tpcds" / "queries"
+    query_file = sql_queries_path / f"{str(query_index).zfill(2)}.sql"
```
Comment (jaychia) [inline benchmarking/tpcds/__main__.py:26]: nit: Python supports formatting of ints, something like `f"{query_index:02}"` I think

File: benchmarking/tpcds/__main__.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    with open(query_file) as f:
+        query = f.read()
+
+    start = datetime.now()
+
+    try:
+        daft.sql(query, catalog=catalog).collect()
```
Comment (jaychia) [inline benchmarking/tpcds/__main__.py:33]: Let's do a `.explain` before the `.collect()`? Also good to at least print the result after collecting.

File: benchmarking/tpcds/__main__.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    catalog = generate_tpcds_catalog(basedir)
+    for query_index in range(1, 100):
+        duration, was_successful, error_msg = run_query(query_index, catalog)
+        if was_successful:
+            successes[query_index] = duration
+        else:
+            failures[query_index] = (duration, error_msg)
+
+    return successes, failures
```
Comment (jaychia) [inline benchmarking/tpcds/__main__.py:57]: It would be simpler if we just do:

```
results = []
...
results.append(Result(duration=..., error_msg=None, ...))
```

This will make parsing of results into CSVs later easier

File: benchmarking/tpcds/datagen.py [hunk 6]
```diff
@@ -1,39 +1,52 @@
...
+
+    basedir.mkdir(parents=True, exist_ok=True)
+    db = duckdb.connect(database=basedir / "tpcds.db")
     db.sql(f"call dsdgen(sf = {scale_factor})")
     for item in db.sql("show tables").fetchall():
         tbl = item[0]
-        print(f"Exporting {tbl} to {basedir}/{tbl}.parquet")
-        db.sql(f"COPY {tbl} TO '{basedir}/{tbl}.parquet'")
+        parquet_file = basedir / f"{tbl}.parquet"
+        print(f"Exporting {tbl} to {parquet_file}")
```
Comment (jaychia) [inline benchmarking/tpcds/datagen.py:25]: nit: switch to logging as well

**Review by jaychia — APPROVED** [review]

File: benchmarking/tpcds/__main__.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    assert tpcds_gen_folder.exists()
+    assert args.scale_factor > 0
+
+    basedir = tpcds_gen_folder / str(args.scale_factor)
+    datagen.gen_tpcds(basedir=basedir, scale_factor=args.scale_factor)
+    run_benchmarks(basedir)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (universalmind303) [inline benchmarking/tpcds/__main__.py:83]: why does this benchmarking script also generate the data? shouldn't we only generate it if it doesn't already exist. 
## uses existing tpcds dataset
```sh
python -m benchmarking.tpcds \
  --scale-factor 1
  --tpcds-folder data/existing_tpcds # existing path
```

## generates new tpcds dataset
```sh
python -m benchmarking.tpcds \
  --scale-factor 1
  --tpcds-folder data/tpcds # path does not exist
```

**Comment by universalmind303** [comment]
> # Usage
> 
> ```shell
> python -m benchmarking.tpcds \
>   --scale-factor 1
>   --tpcds-gen-folder data/tpcds
> ```
> 
> The `--scale-factor` and `--tpcds-gen-folder` variables are _not_ needed (default to `0.01` and `data/tpcds`, respectively).

not necessary, but IMO it'd be nice to add this to the `Makefile` similar to our `dsdgen` make command:  `make dsdgen SCALE_FACTOR=1 OUTPUT_DIR=data/tpcds`


something like: 

`make bench_tpcds SF=1 TPCDS_DIR=data/tpcds`

File: benchmarking/tpcds/__main__.py (hunk 7 above)
Comment (raunakab) [reply benchmarking/tpcds/__main__.py:83]: The `gen_tpcds` function currently checks to see if the `basedir` you passed in exists. If it does, it will not do anything. If it doesn't, then only will it generate the tpcds data.

File: benchmarking/tpcds/__main__.py (hunk 1 above)
Comment (raunakab) [reply benchmarking/tpcds/__main__.py:14]: Updated.

File: benchmarking/tpcds/datagen.py (hunk 6 above)
Comment (raunakab) [reply benchmarking/tpcds/datagen.py:25]: Updated.

File: benchmarking/tpcds/__main__.py (hunk 3 above)
Comment (raunakab) [reply benchmarking/tpcds/__main__.py]: Updated.

File: benchmarking/tpcds/__main__.py (hunk 4 above)
Comment (raunakab) [reply benchmarking/tpcds/__main__.py]: Updated.

File: benchmarking/tpcds/__main__.py (hunk 5 above)
Comment (raunakab) [reply benchmarking/tpcds/__main__.py]: Updated.

File: benchmarking/tpcds/__main__.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+
+from ..tpch import __main__ as tpch
+from ..tpch import ray_job_runner
+from . import datagen, helpers
+
+logger = logging.getLogger(__name__)
+
+SQL_QUERIES_PATH = Path(__file__).parent / "queries"
+
+# Results = tuple[dict[int, float], dict[int, str]]
```
Comment (raunakab) [inline benchmarking/tpcds/__main__.py:21]: Remove

Comment (raunakab) [reply benchmarking/tpcds/__main__.py]: Done.
